src/main/java/de/tum/in/www1/artemis/domain/Exercise.java

```diff
@@ -507,6 +508,7 @@ public void setExerciseHints(Set<ExerciseHint> exerciseHints) {
     // jhipster-needle-entity-add-getters-setters - JHipster will add getters and setters here, do not remove
 
     public Boolean isEnded() {
+        assert isDateUsageAllowed() : "due date is not defined for exam exercises";
```

Comment:
I am not a big fan of such assertions. We won't see the assertion errors on production as they are hidden in the log. I think it would be better to directly send error messages to Sentry

Reply:
One improvement would be to add JavaDoc that explicitly says if the method can be used in a specific case.
Another one would be to give an approximation in case of exam exercises. They typically include the exercise group and exam. In case those related objects are available, we could use the end date of the exam or the start of the exam to give an answer to this method.

In the JavaDoc, we could strongly advice to use the examDateService instead